File: L2ScriptMaker/NpcDataPassiveSkillFix.vb
Public Class NpcDataPassiveSkillFix

    'npcdata.txt:	skill_list={@s_race_demonic;@s_evil_attack;@s_npc_resist_unholy3}
    'npcgrp.txt:	property_list={4298;4278;4333}
    '[s_power_strike11]	=	769

    Private Sub StartButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles StartButton.Click

        Dim NpcGrpFile As String        ' Name/path of npcgrp.txt file
        Dim aNpcSkill(30000) As String   ' Skilllist on each mob from npcgrp.txt file
        Dim SkillPchFile As String        ' Name/path of skillpch.txt file
        Dim aSkillPch(10000) As String   ' Skilllist on each mob from npcgrp.txt file
        Dim NpcDataFile As String        ' Name/path of npcgrp.txt file

        Dim TempStr As String, TempStr2 As String
        Dim TempData() As String
        Dim iTemp As Integer


        OpenFileDialog.Filter = "Lineage client text file (npcgrp.txt)|npcgrp.txt|All files|*.*"
        If OpenFileDialog.ShowDialog = Windows.Forms.DialogResult.Cancel Then Exit Sub
        NpcGrpFile = OpenFileDialog.FileName

        OpenFileDialog.Filter = "Lineage server script file (skill_pch.txt)|skill_pch.txt|All files|*.*"
        If OpenFileDialog.ShowDialog = Windows.Forms.DialogResult.Cancel Then Exit Sub
        SkillPchFile = OpenFileDialog.FileName

        OpenFileDialog.Filter = "Lineage server script file (npcdata.txt)|npcdata.txt|All files|*.*"
        If OpenFileDialog.ShowDialog = Windows.Forms.DialogResult.Cancel Then Exit Sub
        NpcDataFile = OpenFileDialog.FileName

        ' ---- Loading NpcSkills ---- 
        Dim iNpcGrpFile As New System.IO.StreamReader(NpcGrpFile, System.Text.Encoding.Default, True)
        ProgressBar.Maximum = CInt(iNpcGrpFile.BaseStream.Length)
        ProgressBar.Value = 0
        While iNpcGrpFile.EndOfStream <> True

            TempStr = iNpcGrpFile.ReadLine
            If TempStr <> Nothing And TempStr.StartsWith("//") = False Then
                TempStr2 = Libraries.GetNeedParamFromStr(TempStr, "property_list")
                If TempStr2 <> "{0}" Then
                    aNpcSkill(CInt(Libraries.GetNeedParamFromStr(TempStr, "npc_id"))) = Libraries.GetNeedParamFromStr(TempStr, "property_list").Replace("{", "").Replace("}", "")
                End If
            End If
            ProgressBar.Value = CInt(iNpcGrpFile.BaseStream.Position * 100 / iNpcGrpFile.BaseStream.Length)
            Me.Update()

        End While
        iNpcGrpFile.Close()
        ProgressBar.Value = 0
        ' End of Loading NpcSkills

        ' ---- Loading Skills ---- 
        '[s_power_strike11]	=	769
        Dim iSkillPchFile As New System.IO.StreamReader(SkillPchFile, System.Text.Encoding.Default, True)
        ProgressBar.Maximum = CInt(iSkillPchFile.BaseStream.Length)
        ProgressBar.Value = 0
        While iSkillPchFile.EndOfStream <> True

            TempStr = Replace(iSkillPchFile.ReadLine, Chr(9), " ")
            If TempStr <> Nothing And TempStr.StartsWith("//") = False Then

                ' tabs and spaces correction
                TempStr = Replace(TempStr, " ", "")
                TempData = TempStr.Split(CChar("="))

                iTemp = CInt(Fix(CInt(TempData(1)) / 256))
                If iTemp > aSkillPch.Length Then
                    Array.Resize(aSkillPch, iTemp)
                End If

                'Fix with LvlSklBox
                If LvlSklBox.Checked = True And aSkillPch(iTemp) <> Nothing Then
                    aSkillPch(iTemp) = ClearDigit(aSkillPch(iTemp))
                Else
                    aSkillPch(iTemp) = TempData(0).Replace("[", "").Replace("]", "")
                End If

            End If
            ProgressBar.Value = CInt(iSkillPchFile.BaseStream.Position * 100 / iSkillPchFile.BaseStream.Length)
            Me.Update()

        End While
        iSkillPchFile.Close()
        ProgressBar.Value = 0
        ' End of Loading Skills

        ' ---- Main module ----

        Dim iNpcDataFile As New System.IO.StreamReader(NpcDataFile, System.Text.Encoding.Default, True)
        Dim oNpcDataFile As New System.IO.StreamWriter(NpcDataFile & ".fixed.txt", False, System.Text.Encoding.Unicode, 1)
        Dim outFile As New System.IO.StreamWriter(NpcDataFile + "_scriptCheck.log", True, System.Text.Encoding.Unicode, 1)

        ProgressBar.Maximum = CInt(iNpcDataFile.BaseStream.Length)
        ProgressBar.Value = 0

        'npc_begin	warrior	1	[gremlin]	level=1
        'npcdata.txt:	skill_list={@s_race_demonic;@s_evil_attack;@s_npc_resist_unholy3}
        'npcgrp.txt:	property_list={4298;4278;4333}  aNpcSkill
        'aSkillPch

        Dim NpcId As Integer
        Dim OldSkillList As String, NewSkillList As String
        Dim aOldSkill() As String = {}
        Dim aNewSkill() As String = {}

        outFile.WriteLine(vbNewLine & "L2ScriptMaker. Npcdata skill_list fixer.")
        outFile.WriteLine(Now & " Start" & vbNewLine)

        While iNpcDataFile.EndOfStream <> True

            TempStr = Replace(iNpcDataFile.ReadLine, Chr(9), " ")
            If TempStr <> Nothing And TempStr.StartsWith("//") = False Then

                ' tabs and spaces correction
                While InStr(TempStr, "  ") <> 0
                    TempStr = Replace(TempStr, "  ", " ")
                End While
                TempData = TempStr.Split(Chr(32))

                'Preparing
                Array.Clear(aNewSkill, 0, aNewSkill.Length)
                Array.Clear(aOldSkill, 0, aOldSkill.Length)
                OldSkillList = "" : NewSkillList = ""

                NpcId = CInt(TempData(2))

                OldSkillList = Libraries.GetNeedParamFromStr(TempStr, "skill_list")

                If aNpcSkill(NpcId) <> Nothing Then

                    aNewSkill = aNpcSkill(NpcId).Split(CChar(";"))

                    NewSkillList = "{"
                    For iTemp = 0 To aNewSkill.Length - 1
                        'npcdata.txt:	skill_list={@s_race_demonic;@s_evil_attack;@s_npc_resist_unholy3}
                        If iTemp > 0 And iTemp <= (aNewSkill.Length - 1) Then
                            NewSkillList += ";"
                        End If

                        ' ----- Fix with LvlSklBox ----
                        Dim sTemp1 As String = ""
                        If LvlSklBox.Checked = True Then
                            aOldSkill = OldSkillList.Replace("{", "").Replace("}", "").Replace("@", "").Split(CChar(";"))

                            Dim iTemp1 As Integer, iTemp2 As Integer
                            iTemp1 = CInt(aNewSkill(iTemp))
                            sTemp1 = ClearDigit(aSkillPch(iTemp1))
                            For iTemp2 = 0 To aOldSkill.Length - 1
                                If InStr(aOldSkill(iTemp2), sTemp1) <> 0 Then
                                    sTemp1 = aOldSkill(iTemp2)
                                    Exit For
                                End If
                            Next
                        Else
                            sTemp1 = aSkillPch(CInt(aNewSkill(iTemp)))
                        End If
                        '---- end of fix ---------

                        NewSkillList += "@" & sTemp1

                    Next
                    NewSkillList += "}"

                Else
                    NewSkillList = "{}"
                End If

                ' ---- @s_full_magic_defence fix ----
                If SFullMagicDefBox.Checked = True Then
                    For iTemp = 0 To SkillIgnorListBox.Lines.Length - 1
                        If OldSkillList.Contains(SkillIgnorListBox.Lines(iTemp)) = True And NewSkillList.Contains(SkillIgnorListBox.Lines(iTemp)) = False Then

                            ' fix for s_npc_high_level_1 and s_npc_high_level_10
                            If (iTemp = SkillIgnorListBox.Lines.Length - 1) And OldSkillList.Contains("@s_npc_high_level_10") = True Then
                                Exit For
                            End If

                            If NewSkillList <> "{}" Then NewSkillList = NewSkillList.Replace("}", ";}")
                            NewSkillList = NewSkillList.Replace("}", SkillIgnorListBox.Lines(iTemp) & "}")
                        End If
                    Next
                End If
                ' end of @s_full_magic_defence fix

                If OldSkillList <> NewSkillList Then
                    TempStr = TempStr.Replace(OldSkillList, NewSkillList).Replace(Chr(32), Chr(9))
                    outFile.WriteLine("Fixing npc: " & TempData(0) & vbTab & TempData(1) & vbTab & TempData(2) & vbTab & TempData(3))
                    outFile.WriteLine(OldSkillList & vbNewLine & "-->" & vbNewLine & NewSkillList & vbNewLine)
                End If

            End If
            TempStr = TempStr.Replace(Chr(32), Chr(9))
            oNpcDataFile.WriteLine(TempStr)

            ProgressBar.Value = CInt(iNpcDataFile.BaseStream.Position)
            ProgressBar.Update()

        End While

        outFile.WriteLine(Now & " End")
        outFile.Close()
        oNpcDataFile.Close()
        ProgressBar.Value = 0

        MessageBox.Show("Done.")


    End Sub

    Private Sub QuitButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles QuitButton.Click
        Me.Close()
    End Sub

    Private Function ClearDigit(ByRef Name As String) As String

        'ClearDigit = Name.Replace("1", "").Replace("2", "").Replace("3", "").Replace("4", "").Replace("5", "").Replace("6", "").Replace("7", "").Replace("8", "").Replace("9", "").Replace("0", "")
        'Dim cTemp() As Char = {CChar("1"), CChar("2"), CChar("3"), CChar("4"), CChar("5"), CChar("6"), CChar("7"), CChar("8"), CChar("9"), CChar("0")}
        Dim iTemp As Integer
        For itemp = Name.Length - 1 To 0 Step -1
            If Char.IsDigit(Name(itemp)) = True Then
                Name = Name.Remove(iTemp, 1)
            Else
                Exit For
            End If
        Next
        ClearDigit = Name

    End Function
End Class

File: L2ScriptMaker/Modules/Npc/NpcDataPassiveSkillFix.cs
using L2ScriptMaker.Extensions;
using L2ScriptMaker.Extensions.VbCompatibleHelper;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace L2ScriptMaker.Modules.Npc
{
	public partial class NpcDataPassiveSkillFix : Form
	{
		public NpcDataPassiveSkillFix()
		{
			InitializeComponent();
		}


		// npcdata.txt:	skill_list={@s_race_demonic;@s_evil_attack;@s_npc_resist_unholy3}
		// npcgrp.txt:	property_list={4298;4278;4333}
		// [s_power_strike11]	=	769

		private void StartButton_Click(object sender, EventArgs e)
		{
			string NpcGrpFile;        // Name/path of npcgrp.txt file
			var aNpcSkill = new string[30001];   // Skilllist on each mob from npcgrp.txt file
			string SkillPchFile;        // Name/path of skillpch.txt file
			var aSkillPch = new string[10001];   // Skilllist on each mob from npcgrp.txt file
			string NpcDataFile;        // Name/path of npcgrp.txt file
			string TempStr;
			string TempStr2;
			string[] TempData;
			int iTemp;


			OpenFileDialog.Filter = "Lineage client text file (npcgrp.txt)|npcgrp.txt|All files|*.*";
			if (OpenFileDialog.ShowDialog() == DialogResult.Cancel)
				return;
			NpcGrpFile = OpenFileDialog.FileName;

			OpenFileDialog.Filter = "Lineage server script file (skill_pch.txt)|skill_pch.txt|All files|*.*";
			if (OpenFileDialog.ShowDialog() == DialogResult.Cancel)
				return;
			SkillPchFile = OpenFileDialog.FileName;

			OpenFileDialog.Filter = "Lineage server script file (npcdata.txt)|npcdata.txt|All files|*.*";
			if (OpenFileDialog.ShowDialog() == DialogResult.Cancel)
				return;
			NpcDataFile = OpenFileDialog.FileName;

			// ---- Loading NpcSkills ---- 
			var iNpcGrpFile = new System.IO.StreamReader(NpcGrpFile, System.Text.Encoding.Default, true);
			ProgressBar.Maximum = Conversions.ToInteger(iNpcGrpFile.BaseStream.Length);
			ProgressBar.Value = 0;
			while (iNpcGrpFile.EndOfStream != true)
			{
				TempStr = iNpcGrpFile.ReadLine();
				if (TempStr != null & TempStr.StartsWith("//") == false)
				{
					TempStr2 = Libraries.GetNeedParamFromStr(TempStr, "property_list");
					if ((TempStr2 ?? "") != "{0}")
						aNpcSkill[Conversions.ToInteger(Libraries.GetNeedParamFromStr(TempStr, "npc_id"))] = Libraries.GetNeedParamFromStr(TempStr, "property_list").Replace("{", "").Replace("}", "");
				}
				ProgressBar.Value = Conversions.ToInteger(iNpcGrpFile.BaseStream.Position * 100 / (double)iNpcGrpFile.BaseStream.Length);
				this.Update();
			}
			iNpcGrpFile.Close();
			ProgressBar.Value = 0;
			// End of Loading NpcSkills

			// ---- Loading Skills ---- 
			// [s_power_strike11]	=	769
			var iSkillPchFile = new System.IO.StreamReader(SkillPchFile, System.Text.Encoding.Default, true);
			ProgressBar.Maximum = Conversions.ToInteger(iSkillPchFile.BaseStream.Length);
			ProgressBar.Value = 0;
			while (iSkillPchFile.EndOfStream != true)
			{
				TempStr = Strings.Replace(iSkillPchFile.ReadLine(), Conversions.ToString((char)9), " ");
				if (TempStr != null & TempStr.StartsWith("//") == false)
				{

					// tabs and spaces correction
					TempStr = Strings.Replace(TempStr, " ", "");
					TempData = TempStr.Split(Conversions.ToChar("="));

					iTemp = Conversions.ToInteger(Conversions.Fix(Conversions.ToInteger(TempData[1]) / (double)256));
					if (iTemp > aSkillPch.Length)
						Array.Resize(ref aSkillPch, iTemp);

					// Fix with LvlSklBox
					if (LvlSklBox.Checked == true & aSkillPch[iTemp] != null)
					{
						var argName = aSkillPch[iTemp];
						aSkillPch[iTemp] = ClearDigit(ref argName);
					}
					else
						aSkillPch[iTemp] = TempData[0].Replace("[", "").Replace("]", "");
				}
				ProgressBar.Value = Conversions.ToInteger(iSkillPchFile.BaseStream.Position * 100 / (double)iSkillPchFile.BaseStream.Length);
				this.Update();
			}
			iSkillPchFile.Close();
			ProgressBar.Value = 0;
			// End of Loading Skills

			// ---- Main module ----

			var iNpcDataFile = new System.IO.StreamReader(NpcDataFile, System.Text.Encoding.Default, true);
			var oNpcDataFile = new System.IO.StreamWriter(NpcDataFile + ".fixed.txt", false, System.Text.Encoding.Unicode, 1);
			var outFile = new System.IO.StreamWriter(NpcDataFile + "_scriptCheck.log", true, System.Text.Encoding.Unicode, 1);

			ProgressBar.Maximum = Conversions.ToInteger(iNpcDataFile.BaseStream.Length);
			ProgressBar.Value = 0;

			// npc_begin	warrior	1	[gremlin]	level=1
			// npcdata.txt:	skill_list={@s_race_demonic;@s_evil_attack;@s_npc_resist_unholy3}
			// npcgrp.txt:	property_list={4298;4278;4333}  aNpcSkill
			// aSkillPch

			int NpcId;
			string OldSkillList;
			string NewSkillList;
			var aOldSkill = new string[0];
			var aNewSkill = new string[0];

			outFile.WriteLine(Constants.vbNewLine + "L2ScriptMaker. Npcdata skill_list fixer.");
			outFile.WriteLine(Conversions.ToString(DateAndTime.Now) + " Start" + Constants.vbNewLine);

			while (iNpcDataFile.EndOfStream != true)
			{
				TempStr = Strings.Replace(iNpcDataFile.ReadLine(), Conversions.ToString((char)9), " ");
				if (TempStr != null & TempStr.StartsWith("//") == false)
				{

					// tabs and spaces correction
					while (Strings.InStr(TempStr, "  ") != 0)
						TempStr = Strings.Replace(TempStr, "  ", " ");
					TempData = TempStr.Split((char)32);

					// Preparing
					Array.Clear(aNewSkill, 0, aNewSkill.Length);
					Array.Clear(aOldSkill, 0, aOldSkill.Length);
					OldSkillList = ""; NewSkillList = "";

					NpcId = Conversions.ToInteger(TempData[2]);

					OldSkillList = Libraries.GetNeedParamFromStr(TempStr, "skill_list");

					if (aNpcSkill[NpcId] != null)
					{
						aNewSkill = aNpcSkill[NpcId].Split(Conversions.ToChar(";"));

						NewSkillList = "{";
						var loopTo = aNewSkill.Length - 1;
						for (iTemp = 0; iTemp <= loopTo; iTemp++)
						{
							// npcdata.txt:	skill_list={@s_race_demonic;@s_evil_attack;@s_npc_resist_unholy3}
							if (iTemp > 0 & iTemp <= aNewSkill.Length - 1)
								NewSkillList += ";";

							// ----- Fix with LvlSklBox ----
							string sTemp1 = "";
							if (LvlSklBox.Checked == true)
							{
								aOldSkill = OldSkillList.Replace("{", "").Replace("}", "").Replace("@", "").Split(Conversions.ToChar(";"));
								int iTemp1;
								int iTemp2;
								iTemp1 = Conversions.ToInteger(aNewSkill[iTemp]);
								var argName1 = aSkillPch[iTemp1];
								sTemp1 = ClearDigit(ref argName1);
								var loopTo1 = aOldSkill.Length - 1;
								for (iTemp2 = 0; iTemp2 <= loopTo1; iTemp2++)
								{
									if (Strings.InStr(aOldSkill[iTemp2], sTemp1) != 0)
									{
										sTemp1 = aOldSkill[iTemp2];
										break;
									}
								}
							}
							else
								sTemp1 = aSkillPch[Conversions.ToInteger(aNewSkill[iTemp])];
							// ---- end of fix ---------

							NewSkillList += "@" + sTemp1;
						}
						NewSkillList += "}";
					}
					else
						NewSkillList = "{}";

					// ---- @s_full_magic_defence fix ----
					if (SFullMagicDefBox.Checked == true)
					{
						var loopTo2 = SkillIgnorListBox.Lines.Length - 1;
						for (iTemp = 0; iTemp <= loopTo2; iTemp++)
						{
							if (OldSkillList.Contains(SkillIgnorListBox.Lines[iTemp]) == true & NewSkillList.Contains(SkillIgnorListBox.Lines[iTemp]) == false)
							{

								// fix for s_npc_high_level_1 and s_npc_high_level_10
								if (iTemp == SkillIgnorListBox.Lines.Length - 1 & OldSkillList.Contains("@s_npc_high_level_10") == true)
									break;

								if ((NewSkillList ?? "") != "{}")
									NewSkillList = NewSkillList.Replace("}", ";}");
								NewSkillList = NewSkillList.Replace("}", SkillIgnorListBox.Lines[iTemp] + "}");
							}
						}
					}
					// end of @s_full_magic_defence fix

					if ((OldSkillList ?? "") != (NewSkillList ?? ""))
					{
						TempStr = TempStr.Replace(OldSkillList, NewSkillList).Replace((char)32, (char)9);
						outFile.WriteLine("Fixing npc: " + TempData[0] + Constants.vbTab + TempData[1] + Constants.vbTab + TempData[2] + Constants.vbTab + TempData[3]);
						outFile.WriteLine(OldSkillList + Constants.vbNewLine + "-->" + Constants.vbNewLine + NewSkillList + Constants.vbNewLine);
					}
				}
				TempStr = TempStr.Replace((char)32, (char)9);
				oNpcDataFile.WriteLine(TempStr);

				ProgressBar.Value = Conversions.ToInteger(iNpcDataFile.BaseStream.Position);
				ProgressBar.Update();
			}

			outFile.WriteLine(Conversions.ToString(DateAndTime.Now) + " End");
			outFile.Close();
			oNpcDataFile.Close();
			ProgressBar.Value = 0;

			MessageBox.Show("Done.");
		}

		private void QuitButton_Click(object sender, EventArgs e)
		{
			this.Close();
		}

		private string ClearDigit(ref string Name)
		{
			string ClearDigitRet = default(string);

			// ClearDigit = Name.Replace("1", "").Replace("2", "").Replace("3", "").Replace("4", "").Replace("5", "").Replace("6", "").Replace("7", "").Replace("8", "").Replace("9", "").Replace("0", "")
			// Dim cTemp() As Char = {CChar("1"), CChar("2"), CChar("3"), CChar("4"), CChar("5"), CChar("6"), CChar("7"), CChar("8"), CChar("9"), CChar("0")}
			int iTemp;
			for (iTemp = Name.Length - 1; iTemp >= 0; iTemp += -1)
			{
				if (char.IsDigit(Name[iTemp]) == true)
					Name = Name.Remove(iTemp, 1);
				else
					break;
			}
			ClearDigitRet = Name;
			return ClearDigitRet;
		}
	}
}
